rig/routing_table/minimise.py

```diff
@@ -0,0 +1,112 @@
+import collections
+from rig.routing_table import MinimisationFailedError
+from rig.routing_table.remove_default_routes import minimise as \
+    remove_default_entries
+from rig.routing_table.ordered_covering import minimise as ordered_covering
+from six import iteritems
+
+
+def minimise(routing_tables, target_lengths,
+             methods=(remove_default_entries, ordered_covering)):
+    """
+
+    Parameters
+    ----------
+    routing_tables : {(x, y): [RoutingTableEntry, ...], ...}
+        Dictionary mapping co-ordinates to the routing tables associated with
+        those co-ordinates. NOTE: This is the data structure as returned by
+        :py:meth:`~rig.place_and_route.build_full_routing_tables`.
+    target_lengths : int or {(x, y): int or None, ...} or None
+        Maximum length of routing tables. If an integer this is assumed to be
+        the maximum length for any table; if a dictionary then it is assumed to
+        be a mapping from co-ordinate to maximum length (or None); if None then
+        tables will be minimised as far as possible.
+    methods :
+        Iterable of minimisation methods.
+
+    Returns
+    -------
+    {(x, y): [RoutingTableEntry, ...], ...}
+        Minimised routing tables, guaranteed to be smaller than the table sizes
+        specified by `target_lengths`.
+
+    Raises
+    ------
+    MinimisationFailedError
+        If no method can sufficiently minimise a table.
+    """
+    # Coerce the target lengths into the correct forms
+    if not isinstance(target_lengths, dict):
+        lengths = collections.defaultdict(lambda: target_lengths)
+    else:
+        lengths = target_lengths
+
+    # Minimise the routing tables
+    new_tables = dict()
+    for chip, table in iteritems(routing_tables):
+        # Try to minimise the table
+        try:
+            new_table = minimise_table(table, lengths[chip], methods)
+        except MinimisationFailedError as exc:
+            exc.chip = chip
+            raise
+
+        # Store the table if it isn't empty
+        if new_table:
+            new_tables[chip] = new_table
+
+    return new_tables
+
+
+def minimise_table(table, target_length,
+                   methods=(remove_default_entries, ordered_covering)):
+    """Apply different minimisation algorithms to minimise a table below a
```

Comment:
I'd emphasise that this minimises a *single* routing entry just to contrast it with minimise.